### PR TITLE
Properly preserve SymInt input invariant when splitting graphs

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #117406

Fixes https://github.com/pytorch/pytorch/issues/111636
Fixes https://github.com/pytorch/pytorch/issues/108877
Fixes https://github.com/pytorch/pytorch/issues/116956

Inductor has an invariant that every dynamic shape symbol s0, s1, etc. which is referenced by an input tensor must also be passed in explicitly as an argument. It has some capability of reverse engineering symbols if it's obvious how to get them (e.g., if you pass in `arg: f32[s0, 4]` it will know that it can retrieve `s0 = arg.size(0)`) but in full generality it is not always possible to derive this (e.g., if the only mention of s0 is in `arg2: f32[s0 + s1, 4]`).  However, the graph splitter used by optimize_ddp did not respect this invariant. This PR makes it respect it.

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225

### CODE REVIEW

**Comment by ezyang** [comment]
@wconstab PR desc is here, and I actually successfully got it to pass locally lol

File: torch/fx/passes/split_module.py [hunk 1]
```diff
@@ -7,6 +7,8 @@
 from torch.fx._compatibility import compatibility
 from torch.fx.graph_module import GraphModule
 from torch.fx.node import Node
+from torch.fx.experimental.symbolic_shapes import free_symbols
+import sympy
```
Comment (wconstab) [inline torch/fx/passes/split_module.py:11]: hmm, perhaps this workaround would suffice?

if TYPE_CHECKING:
   import sympy

File: torch/fx/passes/split_module.py [hunk 2]
```diff
@@ -7,6 +7,8 @@
 from torch.fx._compatibility import compatibility
 from torch.fx.graph_module import GraphModule
 from torch.fx.node import Node
+from torch.fx.experimental.symbolic_shapes import free_symbols
```
Comment (wconstab) [inline torch/fx/passes/split_module.py:10]: but this part still needs sympy.. so... runtime import?

**Review by wconstab — APPROVED** [review]
lgtm, I think all the CI fails are sympy except 2 infra ones, thanks for the fix!

**Pull request closed without merging**